libmultilabel/linear/metrics.py

```diff
@@ -10,20 +10,94 @@
            'MetricCollection']
 
 
+def _check_top_k(k):
+    if not (isinstance(k, int) and k > 0):
+        raise ValueError('"k" has to be a positive integer')
+
+def _DCG(preds: np.ndarray, target: np.ndarray, k: int = 5) -> np.ndarray:
+    """
+    Compute the discounted cumulative gains (DCG).
+    """
+    # Self-implemented dcg is used here. scikit-learn's implementation has
+    # an average time complexity O(NlogN) as it directly applies quicksort
+    # to preds regardless of k. Here the average time complexity is reduced to
+    # O(N + klogk) by first partitioning off the k largest elements and then
+    # applying quicksort to the subarray.
+    unsorted_top_k_idx = np.argpartition(preds, -k)[:, -k:]
+    unsorted_top_k_scores = np.take_along_axis(
+        preds, unsorted_top_k_idx, -1)
+    sorted_order = np.argsort(-unsorted_top_k_scores)
+    sorted_top_k_idx = np.take_along_axis(
+        unsorted_top_k_idx, sorted_order, -1)
+
```

Comment:
We may need the same subroutine for linear as PR289 needs it as well. Please discuss with Shao-Syuan. This is the only issue left for this PR, so please resolve it soon so we can move on. Thanks